### ISSUE TITLE
Confusion between `done` and `info` in `env.step`, and the correct way we need to detect for when episodes complete.

### ISSUE THREAD

**Issue by DanielTakeshi** [openai/baselines#42]
I ran into an interesting problem today and, while I understand the solution, I'd like to explain it here and inquire about how OpenAI gym and OpenAI baselines are going to handle this going forward.  I'm running gym version 0.9.2 and AtariPy 0.0.20, which is outdated but that's the version where the models were pre-trained for baselines here.

[According to the docs,](https://gym.openai.com/docs) the `env.step` condition returns a `done` parameter which tells us that:

```
done (boolean): whether it's time to reset the environment again. Most (but not all) tasks are divided up into well-defined episodes, and done being True indicates the episode has terminated. (For example, perhaps the pole tipped too far, or you lost your last life.)
```

Note the emphasis on "you lost your last life". This is true, for instance when I run Breakout:
 
```
import gym 
import numpy as np

env = gym.make('Breakout-v0')
obs = env.reset()
done = False
steps = 0 

while not done:
    obs, rew, done, info = env.step(np.random.randint(env.action_space.n))
    steps += 1
    if done:
        print("done == True")
        print("info: {}".format(info))
print("steps: {}".format(steps))
```

The outcome is:

```
[2017-06-29 10:26:00,612] Making new env: Breakout-v0
done == True
info: {'ale.lives': 0}
steps: 271
```

However, the baselines code wraps several monitors around the environment, which results in different semantics of the method. To test, I downloaded the pre-trained Breakout-1 model for Prioritized, Dueling DQN. Then I ran the following command:

```
python -m baselines.deepq.experiments.atari.enjoy --model-dir /tmp/models/model-atari-prior-duel-breakout-1/ --env Breakout --dueling
```

This runs the enjoy script. The only things I changed from the current master branch (version 0778e9f10f4c51417070f24ab12594466443e3a1) are some print statements and removing the render since I was running ssh. You can see the `git diff` here:

```
git diff
diff --git a/baselines/deepq/experiments/atari/enjoy.py b/baselines/deepq/experiments/atari/enjoy.py
index fe482ca..ec5e78e 100644
--- a/baselines/deepq/experiments/atari/enjoy.py
+++ b/baselines/deepq/experiments/atari/enjoy.py
@@ -42,11 +42,13 @@ def play(env, act, stochastic, video_path):
         env, video_path, enabled=video_path is not None)
     obs = env.reset()
     while True:
-        env.unwrapped.render()
+        #env.unwrapped.render()
         video_recorder.capture_frame()
         action = act(np.array(obs)[None], stochastic=stochastic)[0]
         obs, rew, done, info = env.step(action)
         if done:
+            print("done == True")
+            print("info: {}".format(info))
             obs = env.reset()
         if len(info["rewards"]) > num_episodes:
             if len(info["rewards"]) == 1 and video_recorder.enabled:
@@ -56,6 +58,7 @@ def play(env, act, stochastic, video_path):
                 video_recorder.enabled = False
             print(info["rewards"][-1])
             num_episodes = len(info["rewards"])
+            print("we must have finished an episode here now\n")
```

I ran this, but then I saw this output:

```
[2017-06-29 10:22:35,014] Making new env: BreakoutNoFrameskip-v4
done == True
info: {'rewards': [], 'steps': 6845, 'ale.lives': 4}
done == True
info: {'rewards': [], 'steps': 9703, 'ale.lives': 3}
done == True
info: {'rewards': [], 'steps': 10228, 'ale.lives': 2}
done == True
info: {'rewards': [], 'steps': 16350, 'ale.lives': 1}
done == True
info: {'rewards': [], 'steps': 22194, 'ale.lives': 0}
846.0
we must have finished an episode here now

done == True
info: {'rewards': [846.0], 'steps': 24488, 'ale.lives': 4}
done == True
info: {'rewards': [846.0], 'steps': 33442, 'ale.lives': 3}
done == True
info: {'rewards': [846.0], 'steps': 35160, 'ale.lives': 2}
done == True
info: {'rewards': [846.0], 'steps': 37665, 'ale.lives': 1}
done == True
info: {'rewards': [846.0], 'steps': 38732, 'ale.lives': 0}
438.0
we must have finished an episode here now
```

I terminated the run after this, but what happens now is that the `done` semantics have changed and break from the docs. Instead, to detect when an episode finishes, I have to detect when the "rewards" list has increased in size, or when ale.lives is zero. This doesn't seem as elegant as the previous way of just detecting a single `done==True` condition.

In conclusion:

- Detect when an episode finishes with the "rewards" list from `info`, NOT the `done` condition, despite what the documentation says.
- More generally, if the documentation is going to break from default gym, then I think it should clarified somewhere.
- In addition, is there any other set of formal documentation other than the website I linked to earlier, which hasn't changed (as far as I can tell) in about a year and a half.

**Comment by alcinos** [comment]
I think they did that to match what DeepMind did in their original paper. Back then, they considered that an episode is over once you loose your first life (whether that makes sense or not is I think questionnable, but one intuition is that it gives stronger signal to the agent that isn't otherwise punished for death in most games).

So their wrapper basically detects when you loose your first life and set done = True. However as you can see, it is still possible to continue playing since the underlying game is not finished yet. 

To solve you problem, you can simply remove the ``EpisodicLifeEnv`` from the wrapper ( baselines/baselines/common/atari_wrappers_deprecated.py:212)

**Comment by LinZichuan** [comment]
@alcinos Thanks for your explanation! The logger prints five episodic info until live==0, and I found that for each episode their total rewards are same. Is it because they only use the total rewards in first episode?
